File: pq_ai_terminal/plot_waveform.m
%% T536 波形数据读取与绘图脚本
% 用于读取 wave_data.csv 并绘制实时波形图
% 
% 使用方法:
%   1. 确保 MATLAB 路径包含此脚本
%   2. 运行脚本，选择 CSV 文件
%   3. 查看生成的波形图

clear; clc; close all;

%% 配置参数
SAMPLE_RATE = 12800;  % 采样率 Hz
POINTS_PER_CYCLE = 256;  % 每周期采样点数
N_CHANNELS = 7;  % 通道数
CHANNEL_NAMES = {'UA', 'UB', 'UC', 'IA', 'IB', 'IC', 'IZ'};
UNITS = {'V', 'V', 'V', 'A', 'A', 'A', 'A'};

%% 选择数据文件
[filename, filepath] = uigetfile('*.csv', '选择波形数据文件');
if isequal(filename, 0)
    disp('未选择文件，退出。');
    return;
end
fullpath = fullfile(filepath, filename);
fprintf('读取文件: %s\n', fullpath);

%% 解析CSV文件
fid = fopen(fullpath, 'r');
if fid == -1
    error('无法打开文件');
end

cycles = {};
monitor_data = {};
cycle_idx = 0;
monitor_idx = 0;

while ~feof(fid)
    line = fgetl(fid);
    if ~ischar(line)
        break;
    end
    
    % 跳过注释行
    if line(1) == '#'
        % 检查是否是周期头
        if contains(line, '# Cycle')
            cycle_idx = cycle_idx + 1;
            fprintf('解析周期 %d...\n', cycle_idx);
        end
        % 检查是否是监测量头
        if contains(line, '# Monitor')
            monitor_idx = monitor_idx + 1;
        end
        continue;
    end
    
    % 跳过表头
    if startsWith(line, 'point')
        continue;
    end
    
    % 检查是否是监测量数据
    if contains(line, '_RMS') || contains(line, '_Hz') || contains(line, '_kW') || contains(line, '_kvar')
        % 解析监测量
        parts = strsplit(line, ',');
        if length(parts) >= 2
            key = parts{1};
            val = str2double(parts{2});
            if ~isnan(val)
                if monitor_idx > 0
                    monitor_data{monitor_idx}.(key) = val;
                end
            end
        end
        continue;
    end
    
    % 解析波形数据点
    parts = strsplit(line, ',');
    if length(parts) >= 8
        point = str2double(parts{1});
        if cycle_idx > 0 && ~isnan(point)
            data_point = struct();
            data_point.point = point;
            for ch = 1:N_CHANNELS
                data_point.(CHANNEL_NAMES{ch}) = str2double(parts{ch + 1});
            end
            if length(cycles) < cycle_idx
                cycles{cycle_idx} = [];
            end
            cycles{cycle_idx} = [cycles{cycle_idx}; data_point];
        end
    end
end

fclose(fid);

fprintf('解析完成: %d 个周期, %d 个监测量快照\n', cycle_idx, monitor_idx);

%% 提取波形数据矩阵
if cycle_idx > 0 && ~isempty(cycles{1})
    n_cycles = cycle_idx;
    n_points = size(cycles{1}, 1);
    
    % 为每个周期创建矩阵
    wave_matrix = zeros(n_cycles, N_CHANNELS, n_points);
    
    for c = 1:n_cycles
        if length(cycles) >= c && ~isempty(cycles{c})
            for ch = 1:N_CHANNELS
                wave_matrix(c, ch, :) = [cycles{c}.(CHANNEL_NAMES{ch})];
            end
        end
    end
    
    %% 创建时间轴
    time_ms = (0:n_points-1) / SAMPLE_RATE * 1000;  % 毫秒
    
    %% 绘图1: 所有周期叠加显示
    figure('Name', 'T536 实时波形 - 所有周期', 'Position', [100 100 1200 800]);
    
    for ch = 1:N_CHANNELS
        subplot(3, 3, ch);
        hold on;
        for c = 1:min(n_cycles, 5)  % 最多显示5个周期
            plot(time_ms, squeeze(wave_matrix(c, ch, :)), 'b-', 'LineWidth', 1);
        end
        
        % 计算统计值
        data = squeeze(wave_matrix(1, ch, :));
        rms = sqrt(mean(data.^2));
        min_val = min(data);
        max_val = max(data);
        
        title(sprintf('%s (RMS=%.2f%s, Min=%.2f, Max=%.2f)', ...
              CHANNEL_NAMES{ch}, rms, UNITS{ch}, min_val, max_val));
        xlabel('时间 (ms)');
        ylabel(sprintf('电压/电流 (%s)', UNITS{ch}));
        grid on;
        hold off;
    end
    
    sgtitle('T536 实时波形数据 - 多周期叠加', 'FontSize', 14);
    
    %% 绘图2: 最新周期详细显示
    figure('Name', 'T536 实时波形 - 最新周期', 'Position', [100 100 1200 800]);
    
    latest_cycle = n_cycles;
    
    % 电压通道
    subplot(2, 1, 1);
    hold on;
    colors = lines(N_CHANNELS);
    for ch = 1:3  % 电压通道 UA, UB, UC
        plot(time_ms, squeeze(wave_matrix(latest_cycle, ch, :)), ...
             'Color', colors(ch, :), 'LineWidth', 1.5, ...
             'DisplayName', CHANNEL_NAMES{ch});
    end
    title('电压通道波形 (UA, UB, UC)');
    xlabel('时间 (ms)');
    ylabel('电压 (V)');
    legend('Location', 'best');
    grid on;
    hold off;
    
    % 电流通道
    subplot(2, 1, 2);
    hold on;
    for ch = 4:N_CHANNELS  % 电流通道 IA, IB, IC, IZ
        plot(time_ms, squeeze(wave_matrix(latest_cycle, ch, :)), ...
             'Color', colors(ch, :), 'LineWidth', 1.5, ...
             'DisplayName', CHANNEL_NAMES{ch});
    end
    title('电流通道波形 (IA, IB, IC, IZ)');
    xlabel('时间 (ms)');
    ylabel('电流 (A)');
    legend('Location', 'best');
    grid on;
    hold off;
    
    sgtitle('T536 最新周期波形', 'FontSize', 14);
    
    %% 绘图3: 三相电压电流合成图
    figure('Name', 'T536 三相波形合成', 'Position', [100 100 1200 600]);
    
    % 三相电压
    subplot(1, 2, 1);
    hold on;
    phase_colors = {'r', 'g', 'b'};
    for ch = 1:3
        plot(time_ms, squeeze(wave_matrix(latest_cycle, ch, :)), ...
             phase_colors{ch}, 'LineWidth', 2);
    end
    title('三相电压波形');
    xlabel('时间 (ms)');
    ylabel('电压 (V)');
    legend({'UA', 'UB', 'UC'});
    grid on;
    hold off;
    
    % 三相电流
    subplot(1, 2, 2);
    hold on;
    for ch = 4:6
        plot(time_ms, squeeze(wave_matrix(latest_cycle, ch, :)), ...
             phase_colors{ch-3}, 'LineWidth', 2);
    end
    plot(time_ms, squeeze(wave_matrix(latest_cycle, 7, :)), 'k-', 'LineWidth', 1.5);
    title('三相电流 + 零序');
    xlabel('时间 (ms)');
    ylabel('电流 (A)');
    legend({'IA', 'IB', 'IC', 'IZ'});
    grid on;
    hold off;
    
    sgtitle('T536 三相波形分析', 'FontSize', 14);
    
    %% 打印统计信息
    fprintf('\n========== 波形统计信息 ==========\n');
    fprintf('采样率: %d Hz\n', SAMPLE_RATE);
    fprintf('周期数: %d\n', n_cycles);
    fprintf('每周期点数: %d\n', n_points);
    fprintf('\n通道统计 (周期 %d):\n', latest_cycle);
    
    total_power = 0;
    for ch = 1:N_CHANNELS
        data = squeeze(wave_matrix(latest_cycle, ch, :));
        rms = sqrt(mean(data.^2));
        fprintf('  %s: RMS = %8.3f %s, Min = %8.3f, Max = %8.3f\n', ...
                CHANNEL_NAMES{ch}, rms, UNITS{ch}, min(data), max(data));
    end
    
    % 计算总有功功率 (简化: UA*IA + UB*IB + UC*IC)
    p_a = mean(squeeze(wave_matrix(latest_cycle, 1, :)) .* squeeze(wave_matrix(latest_cycle, 4, :)));
    p_b = mean(squeeze(wave_matrix(latest_cycle, 2, :)) .* squeeze(wave_matrix(latest_cycle, 5, :)));
    p_c = mean(squeeze(wave_matrix(latest_cycle, 3, :)) .* squeeze(wave_matrix(latest_cycle, 6, :)));
    total_p = p_a + p_b + p_c;
    fprintf('\n  有功功率: P = %.3f W (Pa=%.2f, Pb=%.2f, Pc=%.2f)\n', total_p, p_a, p_b, p_c);
    
else
    disp('未解析到波形数据！');
end

%% 显示监测量数据
if monitor_idx > 0
    fprintf('\n========== 实时监测量 ==========\n');
    for m = 1:min(monitor_idx, 3)  % 显示前3组
        fprintf('\n监测量快照 %d:\n', m);
        if isfield(monitor_data{m}, 'UA_RMS')
            fprintf('  UA_RMS: %.3f V\n', monitor_data{m}.UA_RMS);
        end
        if isfield(monitor_data{m}, 'UB_RMS')
            fprintf('  UB_RMS: %.3f V\n', monitor_data{m}.UB_RMS);
        end
        if isfield(monitor_data{m}, 'UC_RMS')
            fprintf('  UC_RMS: %.3f V\n', monitor_data{m}.UC_RMS);
        end
        if isfield(monitor_data{m}, 'Frequency_Hz')
            fprintf('  频率: %.3f Hz\n', monitor_data{m}.Frequency_Hz);
        end
    end
end

fprintf('\n========== 完成 ==========\n');

File: pq_ai_terminal/plot_waveform.py
# -*- coding: utf-8 -*-
"""
T536 波形数据绘图工具
用于读取 wave_data.csv 并生成波形图
"""

import numpy as np
import matplotlib.pyplot as plt
import matplotlib
matplotlib.rcParams['font.sans-serif'] = ['SimHei', 'DejaVu Sans']
matplotlib.rcParams['axes.unicode_minus'] = False

# 配置参数
SAMPLE_RATE = 12800  # 采样率 Hz
POINTS_PER_CYCLE = 256  # 每周期采样点数
N_CHANNELS = 7  # 通道数
CHANNEL_NAMES = ['UA', 'UB', 'UC', 'IA', 'IB', 'IC', 'IZ']
UNITS = ['V', 'V', 'V', 'A', 'A', 'A', 'A']


def parse_csv(filepath):
    """解析波形数据CSV文件"""
    cycles = []
    monitors = []
    current_cycle = None
    current_monitor = None

    with open(filepath, 'r') as f:
        for line in f:
            line = line.strip()
            if not line:
                if current_cycle is not None:
                    cycles.append(current_cycle)
                    current_cycle = None
                if current_monitor is not None:
                    monitors.append(current_monitor)
                    current_monitor = None
                continue

            # 跳过注释行
            if line.startswith('#'):
                if '# Cycle' in line:
                    if current_cycle is not None:
                        cycles.append(current_cycle)
                    current_cycle = {'data': []}
                elif '# Monitor' in line:
                    if current_monitor is not None:
                        monitors.append(current_monitor)
                    current_monitor = {}
                continue

            # 跳过表头
            if line.startswith('point'):
                continue

            # 检查是否是监测量数据
            if '_RMS' in line or '_Hz' in line or '_kW' in line or '_kvar' in line:
                parts = line.split(',')
                if len(parts) >= 2 and current_monitor is not None:
                    key = parts[0]
                    try:
                        val = float(parts[1])
                        current_monitor[key] = val
                    except ValueError:
                        pass
                continue

            # 解析波形数据点
            parts = line.split(',')
            if len(parts) >= 8 and current_cycle is not None:
                try:
                    point = float(parts[0])
                    values = [float(parts[i+1]) for i in range(N_CHANNELS)]
                    current_cycle['data'].append([point] + values)
                except ValueError:
                    pass

    # 添加最后一个周期
    if current_cycle is not None:
        cycles.append(current_cycle)
    if current_monitor is not None:
        monitors.append(current_monitor)

    return cycles, monitors


def plot_waveforms(cycles, monitors, output_file='waveform_plot.png'):
    """绘制波形图"""
    if not cycles:
        print("无数据可绘制！")
        return

    n_cycles = len(cycles)
    n_points = len(cycles[0]['data'])
    time_ms = np.linspace(0, n_points / SAMPLE_RATE * 1000, n_points)

    # 提取数据矩阵
    wave_matrix = np.zeros((n_cycles, N_CHANNELS, n_points))
    for c, cycle in enumerate(cycles):
        data = np.array(cycle['data'])
        for ch in range(N_CHANNELS):
            wave_matrix[c, ch, :] = data[:, ch + 1]

    latest = n_cycles - 1

    # 创建图形
    fig = plt.figure(figsize=(16, 12))
    fig.suptitle('T536 实时波形数据采集结果', fontsize=16, fontweight='bold')

    # 子图1: 所有通道叠加
    for ch in range(N_CHANNELS):
        ax = fig.add_subplot(3, 3, ch + 1)
        for c in range(min(n_cycles, 5)):
            ax.plot(time_ms, wave_matrix[c, ch, :], 'b-', linewidth=0.5, alpha=0.7)

        data = wave_matrix[latest, ch, :]
        rms = np.sqrt(np.mean(data**2))
        ax.set_title(f'{CHANNEL_NAMES[ch]} (RMS={rms:.2f}{UNITS[ch]})')
        ax.set_xlabel('时间 (ms)')
        ax.set_ylabel(f'数值 ({UNITS[ch]})')
        ax.grid(True, alpha=0.3)

    plt.tight_layout()

    # 保存图片
    fig.savefig(output_file, dpi=150, bbox_inches='tight')
    print(f"波形图已保存: {output_file}")

    # 创建第二个图形 - 详细波形
    fig2, axes = plt.subplots(2, 2, figsize=(16, 10))
    fig2.suptitle('T536 最新周期波形详细分析', fontsize=14)

    # 电压通道
    ax1 = axes[0, 0]
    colors = ['r', 'g', 'b']
    for ch in range(3):
        ax1.plot(time_ms, wave_matrix[latest, ch, :], colors[ch], linewidth=1.5,
                label=CHANNEL_NAMES[ch])
    ax1.set_title('三相电压波形')
    ax1.set_xlabel('时间 (ms)')
    ax1.set_ylabel('电压 (V)')
    ax1.legend()
    ax1.grid(True, alpha=0.3)

    # 电流通道
    ax2 = axes[0, 1]
    for ch in range(3, 6):
        ax2.plot(time_ms, wave_matrix[latest, ch, :], colors[ch - 3], linewidth=1.5,
                label=CHANNEL_NAMES[ch])
    ax2.plot(time_ms, wave_matrix[latest, 6, :], 'k-', linewidth=1.5, label='IZ')
    ax2.set_title('三相电流 + 零序波形')
    ax2.set_xlabel('时间 (ms)')
    ax2.set_ylabel('电流 (A)')
    ax2.legend()
    ax2.grid(True, alpha=0.3)

    # 电压矢量图
    ax3 = axes[1, 0]
    angles = np.linspace(0, 2 * np.pi, 100)
    for ch in range(3):
        data = wave_matrix[latest, ch, :]
        rms = np.sqrt(np.mean(data**2))
        angle = np.angle(np.mean(data))
        ax3.plot([0, rms * np.cos(angle)], [0, rms * np.sin(angle)],
                colors[ch], linewidth=2, label=f'{CHANNEL_NAMES[ch]} (RMS={rms:.1f}V)')
    ax3.set_title('三相电压矢量图')
    ax3.set_xlabel('实部 (V)')
    ax3.set_ylabel('虚部 (V)')
    ax3.legend()
    ax3.grid(True, alpha=0.3)
    ax3.set_aspect('equal')
    circle = plt.Circle((0, 0), 220, fill=False, linestyle='--', alpha=0.5)
    ax3.add_patch(circle)

    # 功率分析
    ax4 = axes[1, 1]
    p_a = np.mean(wave_matrix[latest, 0, :] * wave_matrix[latest, 3, :])
    p_b = np.mean(wave_matrix[latest, 1, :] * wave_matrix[latest, 4, :])
    p_c = np.mean(wave_matrix[latest, 2, :] * wave_matrix[latest, 5, :])
    total_p = p_a + p_b + p_c

    bars = ax4.bar(['Pa', 'Pb', 'Pc', 'Total'], [p_a, p_b, p_c, total_p],
                   color=['red', 'green', 'blue', 'gray'])
    ax4.set_title('有功功率分析')
    ax4.set_ylabel('功率 (W)')
    ax4.grid(True, alpha=0.3, axis='y')

    for bar in bars:
        height = bar.get_height()
        ax4.text(bar.get_x() + bar.get_width() / 2., height,
                f'{height:.1f}W', ha='center', va='bottom')

    plt.tight_layout()
    fig2.savefig(output_file.replace('.png', '_detail.png'), dpi=150, bbox_inches='tight')
    print(f"详细波形图已保存: {output_file.replace('.png', '_detail.png')}")

    # 打印统计信息
    print("\n========== 波形统计信息 ==========")
    print(f"采样率: {SAMPLE_RATE} Hz")
    print(f"周期数: {n_cycles}")
    print(f"每周期点数: {n_points}")
    print(f"\n通道统计 (周期 {latest + 1}):")

    total_power = 0
    for ch in range(N_CHANNELS):
        data = wave_matrix[latest, ch, :]
        rms = np.sqrt(np.mean(data**2))
        print(f"  {CHANNEL_NAMES[ch]:4s}: RMS = {rms:8.3f} {UNITS[ch]}, "
              f"Min = {np.min(data):8.3f}, Max = {np.max(data):8.3f}")

    print(f"\n  有功功率: P = {total_p:.3f} W (Pa={p_a:.2f}, Pb={p_b:.2f}, Pc={p_c:.2f})")

    if monitors:
        print(f"\n========== 实时监测量 (最新) ==========")
        latest_monitor = monitors[-1]
        for key, val in latest_monitor.items():
            print(f"  {key}: {val:.3f}")

    return wave_matrix, time_ms


def main():
    """主函数"""
    filepath = 'wave_data.csv'

    print(f"读取文件: {filepath}")
    cycles, monitors = parse_csv(filepath)
    print(f"解析完成: {len(cycles)} 个周期, {len(monitors)} 个监测量快照")

    if cycles:
        wave_matrix, time_ms = plot_waveforms(cycles, monitors, 't536_waveform.png')
        print("\n波形分析完成！")
    else:
        print("错误: 未解析到有效数据！")


if __name__ == '__main__':
    main()
